google/cloud/sql/connector/instance_connection_manager.py

```diff
@@ -406,66 +344,134 @@ async def seconds_until_refresh(self) -> int:
-        running _perform_refresh.
+    def _schedule_refresh(self, delay: Optional[int] = None) -> asyncio.Task:
+        """
+        Schedule task to sleep and then perform refresh to get InstanceMetadata.
 
         :rtype: asyncio.Task
-        :returns: A Task representing _get_instance_data.
+        :returns: A Task representing the scheduled _perform_refresh.
         """
 
-        if delay is None:
-            delay = await self.seconds_until_refresh()
-        try:
-            logger.debug("Entering sleep")
-            await asyncio.sleep(delay)
-        except asyncio.CancelledError as e:
-            logger.debug("Schedule refresh task cancelled.")
-            raise e
-        return await self._perform_refresh()
+        async def _refresh_task(
+            self: InstanceConnectionManager, delay: Optional[int]
+        ) -> InstanceMetadata:
+            """
+            A coroutine that sleeps for the specified amount of time before
+            running _perform_refresh.
+            """
+            if delay is None:
+                delay = await self.seconds_until_refresh()
+            refresh_task = self._loop.create_task(self._perform_refresh())
+            try:
+                logger.debug("Entering sleep")
+                await asyncio.sleep(delay)
+                await refresh_task
+            except asyncio.CancelledError as e:
+                logger.debug("Schedule refresh task cancelled.")
+                raise e
+            # bad refresh attempt
+            except Exception as e:
+                logger.exception(
+                    "An error occurred while performing refresh. "
+                    "Scheduling another refresh attempt immediately",
+                    exc_info=e,
+                )
+                # check if current metadata is valid, don't want to replace valid metadata with invalid one
+                instance_data = None
+                try:
+                    instance_data = await self._current
+                except Exception:
+                    # Current result is invalid, no-op
+                    logger.debug("Current instance data is invalid.")
+                # if current metadata is invalid, replace with more recent invalid refresh and schedule another refresh immediately
+                if (
+                    instance_data is None
+                    or instance_data.expiration < datetime.datetime.now()
+                ):
+                    self._current = refresh_task
+                    self._next = self._schedule_refresh(0)
+            # if valid refresh, replace current with valid metadata and schedule next refresh
+            else:
+                self._current = refresh_task
+                # Ephemeral certificate expires in 1 hour, so we schedule a refresh to happen in 55 minutes.
+                self._next = self._schedule_refresh()
+
+            return await refresh_task
```

Comment:
Is it okay to set `refresh_task` to current and next within here? Wonder if there is a way to not need the intermediary task of `refresh_task` and only deal with `scheduled_task`.